pkgs/development/compilers/llvm/14/mlir/default.nix

```diff
@@ -0,0 +1,88 @@
+{ lib, stdenv, llvm_meta
+, monorepoSrc, runCommand
+, cmake
+, libllvm
+, version
+
+# TODO: python bindings
+# TODO: ROCm, CUDA ?
+# TODO: Tests (unit+lit), not integration
+, enableRunners ? stdenv.hostPlatform == stdenv.buildPlatform
+, enableVulkan ? enableRunners
+, vulkan-headers
+, vulkan-loader
+}:
+
+let
+  # List of runners to build, native-only
+  runners = lib.optionals enableRunners ([
+    "cpu-runner" "spirv-cpu-runner"
+  ] ++ lib.optional enableVulkan "vulkan-runner");
+  # List of binaries to build + install
+  # LLVM_{BUILD,INSTALL}_UTILS=ON doesn't seem to work
+  bins = map (n: "mlir-" + n) (runners ++ [
+    "linalg-ods-yaml-gen" "tblgen" # needed for cross
+    "lsp-server" "opt" "pdll" "reduce" "translate" # misc utilities
+  ]);
+in
+stdenv.mkDerivation rec {
+  pname = "mlir";
+  inherit version;
+
+  src = runCommand "${pname}-src-${version}" {} ''
+    mkdir -p "$out"
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/${pname} "$out"
+  '';
+
+  sourceRoot = "${src.name}/${pname}";
+
+  patches = [
+    ./gnu-install-dirs.patch
+  ];
+
+  outputs = [ "out" "lib" "dev" ];
+
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ libllvm ]
+    ++ lib.optionals enableVulkan [ vulkan-headers vulkan-loader ];
+
+  cmakeFlags = [
+    # Documentation suggests packagers may wish to disable, do so until needed
+    "-DMLIR_INSTALL_AGGREGATE_OBJECTS=OFF"
+  ] ++ lib.optionals enableRunners ([
+    "-DMLIR_ENABLE_SPIRV_CPU_RUNNER=ON"
+  ] ++ lib.optional enableVulkan "-DMLIR_ENABLE_VULKAN_RUNNER=ON");
+
+  # Patch around check for being built native (maybe because not built w/LLVM?)
+  postPatch = lib.optionalString enableRunners ''
+    for x in **/CMakeLists.txt; do
+      substituteInPlace "$x" --replace 'if(TARGET ''${LLVM_NATIVE_ARCH})' 'if (1)'
+    done
+  '';
+
+  postBuild = ''
+    make ${lib.concatStringsSep " " bins} -j$NIX_BUILD_CORES -l$NIX_BUILD_CORES
+  '';
+
+  postInstall = ''
+    install -Dm755 -t $out/bin ${lib.concatMapStringsSep " " (x: "bin/${x}") bins}
+
+    mkdir -p $out/share/vim-plugins/
+    cp -r ../utils/vim $out/share/vim-plugins/mlir
+    install -Dt $out/share/emacs/site-lisp ../utils/emacs/mlir-mode.el
```

Comment:
I think the `MLIRConfig.cmake` might also need to be patched. Since the output is split, the `$lib/lib/cmake` folder will be moved to `$dev/lib/cmake`, which might invalidate some variables like [`MLIR_CMAKE_DIR`](https://github.com/llvm/llvm-project/blob/f28c006a5895fc0e329fe15fead81e37457cb1d1/mlir/cmake/modules/MLIRConfig.cmake.in#L9).